clarity/src/vm/analysis/type_checker/mod.rs

```diff
@@ -307,34 +371,311 @@ impl FunctionType {
+        let mut arg_types = Vec::new();
+        for arg in func_args {
+            arg_types.push(self.principal_to_callable_type(arg, 1)?);
+        }
+
+        for (expected_arg, arg_type) in expected_args.iter().zip(arg_types.iter()).into_iter() {
+            inner_type_check_type(
+                db,
+                None,
+                arg_type,
+                &expected_arg.signature,
+                clarity_version,
+                1,
+                &mut LimitedCostTracker::new_free(),
+            )?;
+        }
+        Ok(returns.clone())
+    }
+}
+
+pub fn trait_check_trait_compliance<T: CostTracker>(
+    db: &mut AnalysisDatabase,
+    contract_context: Option<&ContractContext>,
+    actual_trait_identifier: &TraitIdentifier,
+    actual_trait: &BTreeMap<ClarityName, FunctionSignature>,
+    expected_trait_identifier: &TraitIdentifier,
+    expected_trait: &BTreeMap<ClarityName, FunctionSignature>,
+    clarity_version: ClarityVersion,
+    tracker: &mut T,
+) -> CheckResult<()> {
+    // Shortcut for the simple case when the two traits are the same.
+    if actual_trait_identifier == expected_trait_identifier {
+        return Ok(());
+    }
+
+    for (func_name, expected_sig) in expected_trait.iter() {
+        if let Some(func) = actual_trait.get(func_name) {
+            let args_iter = expected_sig.args.iter().zip(func.args.iter());
+            for (expected_type, actual_type) in args_iter {
+                if inner_type_check_type(
+                    db,
+                    contract_context,
+                    actual_type,
+                    expected_type,
+                    clarity_version,
+                    1,
+                    tracker,
+                )
+                .is_err()
+                {
+                    return Err(CheckErrors::IncompatibleTrait(
+                        expected_trait_identifier.clone(),
+                        actual_trait_identifier.clone(),
+                    )
+                    .into());
                 }
-                (expected_type, value) => {
-                    if !expected_type.admits(&value) {
-                        let actual_type = TypeSignature::type_of(&value);
-                        return Err(
-                            CheckErrors::TypeError(expected_type.clone(), actual_type).into()
-                        );
+            }
+            if inner_type_check_type(
+                db,
+                contract_context,
+                &func.returns,
+                &expected_sig.returns,
+                clarity_version,
+                1,
+                tracker,
+            )
+            .is_err()
+            {
+                return Err(CheckErrors::IncompatibleTrait(
+                    expected_trait_identifier.clone(),
+                    actual_trait_identifier.clone(),
+                )
+                .into());
+            }
+        } else {
+            return Err(CheckErrors::IncompatibleTrait(
```

Comment:
It doesn't have to be done in this PR, but it might be helpful for downstream consumers like Clarinet to know a bit more detail about why these traits are incompatible:  it could be because an argument was incompatible, the return type was incompatible, or a required function implementation was missing.

Reply:
Agreed, that would be nice. I opened #3302 to track this.